### PR TITLE
SOC-1657 | Migrate user registration to user-registation service

### PR DESCRIPTION
**Pull request by kvas-damian**
## Links

* https://wikia-inc.atlassian.net/browse/SOC-1657

## Description
* migrated user registration to new user-registation service
* added support for proof-of-work for rate limiting
* migrated facebook registration for response format consistency

## Reviewers

@bkoval @rafalkalinski

### CODE REVIEW

**Comment by rogatty** [comment]
Please run selenium tests on this branch before merging it to make sure they are still up to date.

**Comment by kvas-damian** [comment]
Will do!

File: front/auth/app/common/proof-of-work.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * @param {number} count
+ * @returns {string}
+ */
+function repeatString(string, count) {
+	return new Array(count + 1).join(string);
+}
+
+export default class ProofOfWork {
+	static proof(challenge, bits) {
+		const zeros = Math.floor(bits / 4.0),
```
Comment (rafalkalinski) [inline front/auth/app/common/proof-of-work.js]: (bits / 4)

File: front/auth/app/common/proof-of-work.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+			paddedZeros = repeatString('0', zeros);
+		let counter = 0,
+			digest,
+			out;
+
+		do {
+			digest = CryptoJS.SHA1(`${challenge}${counter.toString(16)}`).toString();
+
+			out = {
+				counter: counter.toString(16),
```
Comment (rafalkalinski) [inline front/auth/app/common/proof-of-work.js]: I think it's worth to add a separate variable for counter.toString(16) value

File: front/auth/app/common/proof-of-work.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ * @param {string} string
+ * @param {number} count
+ * @returns {string}
+ */
+function repeatString(string, count) {
+	return new Array(count + 1).join(string);
+}
+
+export default class ProofOfWork {
+	static proof(challenge, bits) {
```
Comment (rafalkalinski) [inline front/auth/app/common/proof-of-work.js]: maybe you could add a link to some documentation where from this algorithm is.

File: server/app/lib/auth-utils.js [hunk 4]
```diff
@@ -84,6 +84,18 @@ export function getHeliosUrl(path) {
...
 
 /**
+ * @param {string} path
+ * @returns {string}
+ */
+export function getUserRegistrationUrl(path) {
+	return url.format({
+		protocol: 'https',
+		host: localSettings.servicesDomain,
+		pathname: localSettings.userRegistationService.path + path
```
Comment (rafalkalinski) [inline server/app/lib/auth-utils.js]: shouldn't it be more universal? (like checking if there is '?' or '&' in userRegistationService.path or in path)

Comment (kvas-damian) [inline server/app/lib/auth-utils.js]: I don't think we need it. For me it's clear enough that we pass `PATH` as a parameter. https://en.wikipedia.org/wiki/Uniform_Resource_Locator#Syntax

**Comment by bkoval** [comment]
LGTM :+1:

**Comment by bkoval** [comment]
Awesome work!

**Comment by rafalkalinski** [comment]
LGTM!

**Pull request merged**